src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs

```diff
@@ -1000,603 +907,345 @@ private ImmutableArray<Symbol> CreateMembers()
-        {
-            if ((comparison & TypeCompareKind.IgnoreTupleNames) != 0)
+            internal bool EqualsIgnoringTupleUnderlyingType(TupleExtraData? other)
             {
-                if (t2?.IsTupleType == true)
+                if (other is null && this.ElementNames.IsDefault && this.ElementLocations.IsDefault && this.ErrorPositions.IsDefault)
                 {
-                    t2 = t2.TupleUnderlyingType;
+                    return true;
                 }
 
-                return _underlyingType.Equals(t2, comparison, isValueTypeOverrideOpt);
-            }
-
-            return this.Equals(t2 as TupleTypeSymbol, comparison, isValueTypeOverrideOpt);
-        }
+                return other is object
+                    && compareElementNames(other)
+                    && compareElementLocations(other)
+                    && compareErrorPosition(other);
 
-        internal bool Equals(TupleTypeSymbol other)
-        {
-            return Equals(other, TypeCompareKind.ConsiderEverything);
-        }
+                bool compareElementNames(TupleExtraData other)
+                {
+                    if (this.ElementNames.IsDefault && other.ElementNames.IsDefault)
+                    {
+                        return true;
+                    }
 
-        private bool Equals(TupleTypeSymbol other, TypeCompareKind comparison, IReadOnlyDictionary<TypeParameterSymbol, bool> isValueTypeOverrideOpt)
-        {
-            if (ReferenceEquals(this, other))
-            {
-                return true;
-            }
+                    if (this.ElementNames.IsDefault != other.ElementNames.IsDefault)
+                    {
+                        return false;
+                    }
 
-            if ((object)other == null || !other._underlyingType.Equals(_underlyingType, comparison, isValueTypeOverrideOpt))
-            {
-                return false;
-            }
+                    return this.ElementNames.SequenceEqual(other.ElementNames, (n1, n2) => string.Equals(n1, n2, StringComparison.Ordinal));
+                }
 
-            // Make sure field names are the same.
-            if ((comparison & TypeCompareKind.IgnoreTupleNames) == 0)
-            {
-                if (this._elementNames.IsDefault)
+                bool compareElementLocations(TupleExtraData other)
                 {
-                    if (!other._elementNames.IsDefault)
+                    if (this.ElementLocations.IsDefault && other.ElementLocations.IsDefault)
+                    {
+                        return true;
+                    }
+
+                    if (this.ElementLocations.IsDefault != other.ElementLocations.IsDefault)
                     {
                         return false;
                     }
+
+                    return this.ElementLocations.SequenceEqual(other.ElementLocations, (l1, l2) => object.Equals(l1, l2));
                 }
```

Comment:
It looks like all of three of these functions use the default compare. Please consider extracting a `AreEqual<T>(Immutable<T> x, Immutable<T> y)` helper method.